pkg/storj/node.go

```diff
@@ -3,13 +3,174 @@
 
 package storj
 
-import "encoding/hex"
+import (
+	"crypto"
+	"crypto/x509"
+	"encoding/base64"
+	"fmt"
+	"math/bits"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+	"golang.org/x/crypto/sha3"
+	"storj.io/storj/pkg/utils"
+)
+
+// IdentityLength is the number of bytes required to represent a node id
+const IdentityLength = 32
+
+var (
+	ErrNotImplemented = errs.Class("not implemented error")
+	// ErrNodeID is used when something goes wrong with a node id
+	ErrNodeID   = errs.Class("node ID error")
+	// EmptyNodeID is the zero-value for a NodeID
+	EmptyNodeID = NodeID([IdentityLength]byte{})
+)
 
 // NodeID is a unique node identifier
-type NodeID [32]byte
+type NodeID [IdentityLength]byte
+type NodeIDList []NodeID
+
+func NodeIDFromString(s string) (NodeID, error) {
+	idBytes, err := base64.URLEncoding.DecodeString(s)
+	if err != nil {
+		return EmptyNodeID, ErrNodeID.Wrap(err)
+	}
+	return NodeIDFromBytes(idBytes)
+}
+
+func NodeIDsFromBytes(b [][]byte) (ids NodeIDList, err error) {
+	var idErrs []error
+	for _, idBytes := range b {
+		id, err := NodeIDFromBytes(idBytes)
+		if err != nil {
+			idErrs = append(idErrs, err)
+			continue
+		}
+
+		ids = append(ids, id)
+	}
+
+	if err = utils.CombineErrors(idErrs...); err != nil {
+		return nil, err
+	}
+	return ids, nil
+}
 
-// HexString returns NodeID as hex encoded string
-func (id *NodeID) HexString() string { return hex.EncodeToString(id[:]) }
+func NodeIDFromBytes(b []byte) (NodeID, error) {
+	bLen := len(b)
+	if bLen < IdentityLength {
+		return NodeID{}, ErrNodeID.New("not enough bytes to make a node id; have %d, need %d", bLen, IdentityLength)
+	}
+
+	var id NodeID
+	copy(id[:], b[:IdentityLength])
+	return NodeID(id), nil
+}
+
+func NodeIDFromKey(k crypto.PublicKey) (NodeID, error) {
+	kb, err := x509.MarshalPKIXPublicKey(k)
+	if err != nil {
+		return NodeID{}, ErrNodeID.Wrap(err)
+	}
+	hash := make([]byte, IdentityLength)
+	sha3.ShakeSum256(hash, kb)
+	return NodeIDFromBytes(hash)
+}
+
+// String returns NodeID as hex encoded string
+func (id NodeID) String() string {
+	return base64.URLEncoding.EncodeToString(id[:])
+}
 
 // Bytes returns raw bytes of the id
 func (id NodeID) Bytes() []byte { return id[:] }
+
+func (id NodeID) Difficulty() uint16 {
+	idLen := len(id)
+	for i := 1; i < idLen; i++ {
+		b := id[idLen-i]
+
+		if b != 0 {
+			zeroBits := bits.TrailingZeros16(uint16(b))
+			if zeroBits == 16 {
+				zeroBits = 0
+			}
+
+			return uint16((i-1)*8 + zeroBits)
+		}
+	}
+
+	// NB: this should never happen
+	reason := fmt.Sprintf("difficulty matches id hash length: %d; hash (hex): % x", idLen, id)
+	zap.S().Error(reason)
+	panic(reason)
```

Comment:
This should return an error. It isn't unreasonable to get sent a NodeID that is just zeros. If it's annoying to implement right now, then just keep the panic.